core/tracker/tracker.go

```diff
@@ -58,18 +59,29 @@ type Tracker struct {
 	events    map[core.Duty][]event
 	deadliner core.Deadliner
 	quit      chan struct{}
+
+	deadlineFunc   func(duty core.Duty) time.Time
+	reportDutyFunc func(core.Duty, bool, string, string)
 }
 
 // NewTracker returns a new Tracker.
-func NewTracker(deadliner core.Deadliner) *Tracker {
+// TODO(xenowits): implement reportDutyFunc to emit metrics.
+func NewTracker(ctx context.Context, deadlineFunc func(core.Duty) time.Time, reportDutyFunc func(core.Duty, bool, string, string)) (*Tracker, error) {
+	deadliner, err := core.NewDeadliner(ctx, deadlineFunc)
```

Comment:
pass in deadliner, we should not need both deadliner and deadlineFunc, so remove deadlineFunc